includes/data-port/import-tasks/class-sensei-import-associations.php

```diff
@@ -0,0 +1,413 @@
+
+		foreach ( $this->course_lessons as $course_id => $args ) {
+			if ( $this->batch_remaining <= 0 ) {
+				return;
+			}
+
+			$this->handle_course_lessons( $course_id, $args );
+			unset( $this->course_lessons[ $course_id ] );
+		}
+
+		foreach ( $this->lesson_modules as $lesson_id => $args ) {
+			if ( $this->batch_remaining <= 0 ) {
+				return;
+			}
+
+			$this->handle_lesson_module( $lesson_id, $args );
+			unset( $this->lesson_modules[ $lesson_id ] );
+		}
+	}
+
+	/**
+	 * Handle associating a course to a its lessons.
+	 *
+	 * @param int   $course_id Course post ID.
+	 * @param array $args      Arguments from when this task was enqueued.
+	 */
+	private function handle_course_lessons( $course_id, $args ) {
+		$this->batch_remaining--;
+
+		list( $lessons_str, $line_number, $post_title ) = $args;
+
+		$new_lessons = array_unique( Sensei_Data_Port_Utilities::split_list_safely( $lessons_str ) );
+		if ( null === $new_lessons ) {
+			return;
+		}
+
+		$add_warning_helper = function( $message, $code ) use ( $line_number, $course_id, $post_title ) {
+			$model_key = Sensei_Import_Course_Model::MODEL_KEY;
+			$this->get_job()->add_line_warning(
+				$model_key,
+				$line_number,
+				$message,
+				[
+					'line'        => $line_number,
+					'type'        => $model_key,
+					'post_id'     => $course_id,
+					'entry_title' => $post_title,
+					'code'        => $code,
+				]
+			);
+		};
+
+		$current_lessons = $this->get_current_lesson_ids( $course_id );
+		$order_index     = 1;
+		$module_indices  = [];
+		$lesson_order    = [];
+		$all_lessons     = [];
+		foreach ( $new_lessons as $lesson_ref ) {
+			$lesson_id = $this->get_job()->translate_import_id( Sensei_Data_Port_Lesson_Schema::POST_TYPE, $lesson_ref );
+			if ( empty( $lesson_id ) ) {
+				$add_warning_helper(
+					// translators: Placeholder is the reference to a lesson which did not exist.
+					sprintf( __( 'Lesson does not exist: %s.', 'sensei-lms' ), $lesson_ref ),
+					'sensei_data_port_course_lesson_not_found'
+				);
+
+				continue;
+			}
+
+			// Check to see if this lesson was set to multiple courses in the import.
+			$current_lesson_course_id = get_post_meta( $lesson_id, '_lesson_course', true );
+			if (
+				$current_lesson_course_id
+				&& intval( $current_lesson_course_id ) !== intval( $course_id )
+				&& $this->get_job()->was_imported( Sensei_Data_Port_Course_Schema::POST_TYPE, $current_lesson_course_id )
+			) {
+				$add_warning_helper(
+					// translators: Placeholder is the lesson reference (e.g. "id:44").
+					sprintf( __( 'The lesson "%s" can only be associated with one course at a time.', 'sensei-lms' ), $lesson_ref ),
+					'sensei_data_port_lesson_multiple_course'
+				);
+
+				continue;
+			}
+
+			// Add the lesson to the course.
+			update_post_meta( $lesson_id, '_lesson_course', $course_id );
+
+			$module_term_id = false;
+			// Make sure any related module associations are handled.
+			if ( isset( $this->lesson_modules[ $lesson_id ] ) ) {
```

Comment:
This is the main reason we had to put these both in a post-process task. Modules need to be handled after lessons are associated with a course but before lesson order is saved.

Reply:
I would consider having three separate steps (functions) for each association task: 1) link lesson to course, 2) link lesson to module, 3) Order lessons. The main reason for that is that `handle_course_lessons` is a bit complicated ATM and we could avoid calling `handle_lesson_module` from `handle_course_lessons`. This of course would mean that we might need another array to keep track of the order lesson actions. Not sure if it's worth it, WDYT?

Reply:
If I understand correctly, I think we need to keep the handles separate because we may not always be importing courses files with lessons files (and visa versa). I think it is okay for `handle_course_lessons` to call `handle_lesson_module` since it is just doing a task it would do later immediately and we're not putting any extra logic in place (except for having `handle_lesson_module` return the `term_id`) in order for it to be used as a helper.

`handle_course_lessons` is a bit of a long method. I pulled out the bit to get a lesson's module into a new method (9969634c770d713bae615117ac82ba99d2a04731); I renamed the previous method named that to the more descriptive `get_module_for_course`.

Reply:
I was just thinking if we could move all the work done after updating the `_lesson_course` meta to its own method (e.g. handle_lesson_order) which would be executed after `handle_course_lessons` and `handle_lesson_module` in `run`.

> I think it is okay for `handle_course_lessons` to call `handle_lesson_module` since it is just doing a task it would do later immediately

Yep performance-wise the current solution is probably better of what I suggested, I just thought it would make the code more readable. In general I think that the new method improved it though so I don't think that its needed anymore.